libsnark/gadgetlib1/gadgets/hashes/anemoi/anemoi_components.tcc

```diff
@@ -321,32 +321,293 @@ void flystel_prime_field_gadget<ppT, parameters>::generate_r1cs_witness()
+    }
+    return Y;
+}
+
+// rotate left by 1 a vector of elements of type "linear combination of FieldT
+// elements": (x1_0 x1_1 ... x1_{L-1}) -> (x1_1 ... x1_{L-1} x_0)
+template<typename ppT>
+std::vector<linear_combination<libff::Fr<ppT>>> anemoi_vector_left_rotate_by_one(
+    const std::vector<linear_combination<libff::Fr<ppT>>> X_input)
+{
+    if (!((X_input.size() == 2) || (X_input.size() == 3) ||
+          (X_input.size() == 4))) {
+        throw std::invalid_argument("invalid length of input vector");
+    }
+    std::vector<linear_combination<libff::Fr<ppT>>> X = X_input;
+    rotate(X.begin(), X.begin() + 1, X.end());
+    return X;
+}
+
+// rotate left by 1 a vector of elements of type "linear combination of FieldT
+// elements": (x1_0 x1_1 ... x1_{L-1}) -> (x1_1 ... x1_{L-1} x_0)
+template<typename ppT>
+pb_linear_combination_array<libff::Fr<ppT>> anemoi_vector_left_rotate_by_one(
+    const pb_linear_combination_array<libff::Fr<ppT>> X_input)
+{
+    if (!((X_input.size() == 2) || (X_input.size() == 3) ||
+          (X_input.size() == 4))) {
+        throw std::invalid_argument("invalid length of input vector");
+    }
+    pb_linear_combination_array<libff::Fr<ppT>> X = X_input;
+    rotate(X.begin(), X.begin() + 1, X.end());
+    return X;
+}
+
+template<typename ppT, size_t NumStateColumns_L, class parameters>
+anemoi_permutation_round_prime_field_gadget<
+    ppT,
+    NumStateColumns_L,
+    parameters>::
+    anemoi_permutation_round_prime_field_gadget(
+        protoboard<libff::Fr<ppT>> &pb,
+        const std::vector<FieldT> &C,
+        const std::vector<FieldT> &D,
+        const pb_linear_combination_array<FieldT> &X_left,
+        const pb_linear_combination_array<FieldT> &X_right,
+        const pb_variable_array<FieldT> &Y_left,
+        const pb_variable_array<FieldT> &Y_right,
+        const std::string &annotation_prefix)
+    : gadget<libff::Fr<ppT>>(pb, annotation_prefix)
+    , C_const(C)
+    , D_const(D)
+    , X_left_input(X_left)
+    , X_right_input(X_right)
+    , Y_left_output(Y_left)
+    , Y_right_output(Y_right)
+{
+    const libff::Fr<ppT> g = parameters::multiplicative_generator_g;
+    const size_t ncols = NumStateColumns_L;
+
+    // temporary variables (Z_left, Z_right) modified in-place during
+    // the computation from (X_left, X_right) to (Y_left, Y_right)
+    std::vector<linear_combination<libff::Fr<ppT>>> Z_left;
+    std::vector<linear_combination<libff::Fr<ppT>>> Z_right;
+
+    // add constants Z_left[i]+=C[i], Z_right[i]+=D[i]
+    for (size_t i = 0; i < ncols; i++) {
+        Z_left.push_back(X_left[i] + C[i]);
+        Z_right.push_back(X_right[i] + D[i]);
+    }
+
+    if (ncols > 1) {
+        M_matrix = anemoi_permutation_mds<ppT, ncols>(g);
+    } else { // ncols == 1
+        // the MDS matrix for a state with 1 column (L=1) is the same as
+        // for a state with 2 columns (L=2)
+        M_matrix = anemoi_permutation_mds<ppT, 2>(g);
```

Comment:
Can't see where this is used.  Am I missing something?

Reply:
This clause will be activated for `ncols = NumStateColumns_L = 1` e.g. here https://github.com/clearmatics/libsnark/blob/anemoi-hash-r1cs-permutation-round/libsnark/gadgetlib1/gadgets/hashes/anemoi/tests/test_anemoi_gadget.cpp#L308